### PR TITLE
[Merged by Bors] - feat: `Encodable` and `Countable` instances for `Π₀ i, α i` and `α →₀ β`

### LINKED ISSUE #5776: `DFinsupp` and `Finsupp` are countable/encodable
- [ ] Prove that `Π₀ i : ι, α i` is encodable provided that `ι` and
      all `α i` are encodable.
- [ ] Prove that `Π₀ i : ι, α i` is countable provided that `ι` and
      all `α i` are countbale.
- [ ] Prove that `α →₀ β` is encodable if `α` and `β` are encodable.
- [ ] Prove that `α →₀ β` is countable if `α` and `β` are countable.

### PR DESCRIPTION
**Pull request by urkud**
Fixes #5776

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/DFinsupp/Basic.lean [hunk 1]
```diff
@@ -1129,6 +1129,32 @@ theorem eq_mk_support (f : Π₀ i, β i) : f = mk f.support fun i => f i := by
...
+      _ ↔ i ∈ s := by simp⟩
+  left_inv := by
+    rintro ⟨f, rfl⟩
+    ext i
+    simpa using Eq.symm
+  right_inv f := by
+    ext1
+    simp [Subtype.eta]; rfl
+
+/-- Equivalence between all dependent finitely supported functions `f : Π₀ i, β i` and type type
```
Comment (sgouezel) [inline Mathlib/Data/DFinsupp/Basic.lean:1152]:
```suggestion
/-- Equivalence between all dependent finitely supported functions `f : Π₀ i, β i` and type
```

File: Mathlib/Data/DFinsupp/Basic.lean [hunk 2]
```diff
@@ -1129,6 +1129,32 @@ theorem eq_mk_support (f : Π₀ i, β i) : f = mk f.support fun i => f i := by
...
+of pairs `⟨s : Finset ι, f : ∀ i : s, {x : β i // x ≠ 0}⟩`. -/
+@[simps! apply_fst apply_snd_coe]
+def equivSigmaFinsetFun : (Π₀ i, β i) ≃ Σ s : Finset ι, ∀ i : s, {x : β i // x ≠ 0} :=
```
Comment (sgouezel) [inline Mathlib/Data/DFinsupp/Basic.lean:1155]: I am not sure about the naming convention here, but it feels strange that the word `equiv` comes first here while it comes last in the above definition.

**Review by sgouezel — COMMENTED** [review]
bors d+
Thanks!

File: Mathlib/Data/DFinsupp/Basic.lean [hunk 3]
```diff
@@ -1129,6 +1129,32 @@ theorem eq_mk_support (f : Π₀ i, β i) : f = mk f.support fun i => f i := by
...
+    ext i
+    simpa using Eq.symm
+  right_inv f := by
+    ext1
+    simp [Subtype.eta]; rfl
+
+/-- Equivalence between all dependent finitely supported functions `f : Π₀ i, β i` and type
+of pairs `⟨s : Finset ι, f : ∀ i : s, {x : β i // x ≠ 0}⟩`. -/
+@[simps! apply_fst apply_snd_coe]
+def equivSigmaFinsetFun : (Π₀ i, β i) ≃ Σ s : Finset ι, ∀ i : s, {x : β i // x ≠ 0} :=
```
Comment (urkud) [inline Mathlib/Data/DFinsupp/Basic.lean:1155]:
```suggestion
def sigmaFinsetFunEquiv : (Π₀ i, β i) ≃ Σ s : Finset ι, ∀ i : s, {x : β i // x ≠ 0} :=
```

File: Mathlib/Data/DFinsupp/Encodable.lean [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+-/
+
+variable {ι : Type _} {α : ι → Type _} [∀ i, Zero (α i)]
+
+instance [Encodable ι] [∀ i, Encodable (α i)] [∀ i (x : α i), Decidable (x ≠ 0)] :
+    Encodable (Π₀ i, α i) :=
+  letI : DecidableEq ι := Encodable.decidableEqOfEncodable _
+  letI : ∀ s : Finset ι, Encodable (∀ i : s, {x : α i // x ≠ 0}) := fun _ ↦
+    .ofEquiv _ <| .piCongrLeft' _ Encodable.fintypeEquivFin
+  .ofEquiv _ DFinsupp.equivSigmaFinsetFun
```
Comment (urkud) [inline Mathlib/Data/DFinsupp/Encodable.lean:21]:
```suggestion
  .ofEquiv _ DFinsupp.sigmaFinsetFunEquiv
```

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**